build-lib/util/versioner.js

```diff
@@ -0,0 +1,84 @@
+/* global require, module, __dirname */
+
+"use strict";
+
+const Plugin = require('broccoli-plugin');
+const path = require('path');
+const fs = require('fs');
+const git = require('nodegit');
+
+const pkg = require('../../package.json');
+
+function Versioner(inputNodes, options) {
+  const defaultedOptions = options || {};
+
+  Plugin.call(this, inputNodes, {
+    annotation: defaultedOptions.annotation
+  });
+
+  this.options = defaultedOptions;
+}
+
+Versioner.VERSION = pkg.version;
+
+Versioner.prototype = Object.create(Plugin.prototype);
+Versioner.prototype.constructor = Versioner;
+
+Versioner.prototype.build = function () {
+  if (this.constructor.HEAD_COMMIT) {
+    this.versionFiles();
+
+    return null;
+  }
+
+  return git.Repository.open(path.join(__dirname, '..', '..', '.git')).then(repository => {
+    return repository.getHeadCommit().then(commit => {
+      this.constructor.HEAD_COMMIT = commit.id().tostrS().substr(0, 6);
+
+      this.versionFiles();
+    });
+  });
+};
+
+Versioner.prototype.prependComment = function (versionString, buffer) {
+  const versionComment = `/* version: ${versionString} */`;
+
+  return `\n${versionComment}\n${buffer}`;
+};
+
+Versioner.prototype.replaceTemplateString = function (versionString, buffer) {
+  const templateString = this.options.templateString;
+
+  if (templateString) {
+    return buffer.replace(templateString, versionString);
+  }
+
+  return buffer;
+};
+
+Versioner.prototype.versionFiles = function () {
+  const versionString = `v${this.constructor.VERSION}-${this.constructor.HEAD_COMMIT}`;
+
+  const files = this.inputPaths.reduce((fileAcc, dir) => {
+    const list = fs.readdirSync(dir).map(fileName => {
+      return path.join(dir, fileName);
+    });
+
+    return fileAcc.concat(list);
+  }, []).filter(fileName => {
+    return fileName.match(/^.+\.js$/);
```

Comment:
@minasmart, filtering of files to only `.js` files means the `outputPath` would only contain `.js` files. I think we should write all files to the output but version only the `.js` files. Otherwise, the consuming script would have to `mergeTrees` and use the `overwrite` config to ensure the other files are preserved.